presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -226,4 +229,157 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
 
         checkArgument((totalBitLength <= maximumBits), format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
     }
+
+    /**
+     * Searches for and outputs the addresses within certain ranges in each dimension.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of addresses within the given value ranges
+     */
+    public List<Long> zOrderSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        List<Long> addresses = new ArrayList<>();
+
+        List<ZAddressRange> addressRanges = zOrderCondensedSearchCurve(minimumValues, maximumValues);
+
+        Iterator<ZAddressRange> rangeIterator = addressRanges.iterator();
+        while (rangeIterator.hasNext()) {
+            ZAddressRange range = rangeIterator.next();
+            for (long address = range.getMinAddress(); address <= range.getMaxAddress(); address++) {
+                addresses.add(address);
+            }
+        }
+
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    public List<ZAddressRange> zOrderCondensedSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        ZOrderCurveSearcher curveSearcher = new ZOrderCurveSearcher(minimumValues, maximumValues);
+        return curveSearcher.findAddresses();
+    }
+
+    /**
+     * Provides a structure to search for addresses within given value ranges.
+     */
+    private class ZOrderCurveSearcher
+    {
+        private final List<Integer> minimumValues;
+        private final List<Integer> maximumValues;
+        private final int[] dimensions;
+        private LinkedList<ZAddressRange> addressRanges;
+
+        /**
+         * Class constructor specifying the value ranges and initializing <code>dimensions</code>.
+         * <p/>
+         * The <code>dimensions</code> array stores how many values are stored at each level of the Z-curve.
+         */
+        public ZOrderCurveSearcher(List<Integer> minimumValues, List<Integer> maximumValues)
```

Comment:
In our use case, the query may not provide all the min or max values for a dimension. I.e., For table T we have two dimension columns a, b. However, in queries, the query may not use both a and b. e.g., `select ... from ... where a > 0;`. In this case, what is the default min and max value for b dimension? Or we suppose the user should provide default min and max to the library? Let me know your thoughts.